README.md
# hpfilter <img src="man/figures/logo.png" align="right" height="136" />

## An R Implementation of the One- and Two-Sided Hodrick-Prescott Filter

### **hpfilter** is a package that implements the one-sided and two-sided versions of the Hodrick-Prescott filter in R.

The two-sided implementation uses sparse matrices to enhance the efficiency of calculations when using large datasets. Its traditional use has been in macroeconomics, to smooth out the time series of variables such as the GDP. The one-sided version is based on the Kalman filter. An important use of the one-sided HP filter is for the calculation of the credit gap, according to the Basel III methodology for setting the Countercyclical Capital Buffer (CCyB) guide.

The Hodrick-Prescott filter is a technique commonly used to smooth macroeconomic data such as the GDP. It consists in separating short-term, cyclical movements in the data from the long-term trend. Although the HP filter has received numerous critiques, recent studies show that alternative methods do not necessarily perform better. Furthermore, the Basel committee, an international body responsible for setting international guidelines for prudential regulation, has recommended the usage of the one-sided HP filter for calculations pertaining to the analysis of financial cycles.

### [Paper (ResearchGate)](http://dx.doi.org/10.13140/RG.2.2.12251.85282)

### v1.0.0

## About hpfilter

**hpfilter** consists of the following functions:

### hp1 - the One-Sided HP Filter
This function applies the one-sided Hodrick-Prescott filter to the selected data, for the given smoothing parameter value.

In R, simply call the function:

**hp1(_y_, _lambda_, _x_user_, _P_user_, _discard_ )**

_y_	–	a dataframe of size Txn, where “T” is the number of observations for each variable (number of rows) and “n” - the number of variables in the dataframe (number of columns).

_λ_	–	the smoothing parameter; a numeric scalar which takes the default value of 1600, if unspecified by the user.

_x_user_	–	user defined initial values of the state estimate for each variable in y. Takes the form of a 2xn matrix. Since the underlying state vector is 2x1, two values are needed for each variable in y.
By default: if no values are provided, by default, backwards extrapolations based on the first two observations are used.

_P_user_	–	a structural array with n elements, each of which being a 2x2 matrix of initial MSE estimates for each variable in y.
By default: If no values are provided, the default matrix with large variances is used.

_discard_	–	the number of discard periods, expressed as a numeric scalar. The user specified amount of values will be discarded from the start of the sample, resulting in output matrices of size (T-discard)xn.
By default: If no values are provided, the value of 0 is used.

 ### hp2 - the Two-Sided HP Filter
 This function applies the two-sided Hodrick-Prescott filter to the selected data, for the given smoothing parameter value.

**hp2(_y_, _lambda_)**

_y_	–	a dataframe of size Txn, where “T” is the number of observations for each variable (number of rows) and “n” - the number of variables in the dataframe (number of columns).

_λ_	–	the smoothing parameter; a numeric scalar which takes the default value of 1600, if unspecified by the user.

## Technical details

The value of the smoothing parameter should be positive. Furthermore, the code needs at least four observations in order to run. That said, the quality of the estimation will improve with the length of the time series.

## Download and installation of development version

#### Online, from Github:

You can download **hpfilter** directly from Github. To do so, you need to have the **devtools** package installed and loaded. Once you are in **R**, run the following commands:

> install.packages("devtools")
>
> library("devtools")
>
> install_github("alexandrumonahov/hpfilter")

You may face downloading errors from Github if you are behind a firewall or there are https download restrictions. To avoid this, you can try running the following commands:

> options(download.file.method = "libcurl")
>
> options(download.file.method = "wininet")

Once the package is installed, you can run it using the: **library(hpfilter)** command.

## Author details

Alexandru Monahov, 2023



NEWS.md
# hpfilter 1.0.0

This is the initial release.


R/hp2.R
#' Two-Sided HP Filter
#'
#' hp2 applies a standard two-sided Hodrick–Prescott filter using sparse
#' matrices to help reduce the compute time for large datasets. The smoothing
#' parameter should reflect the cyclical duration or frequency of the data.
#'
#' @param y a dataframe of size Txn, where “T” is the number of observations
#'   for each variable (number of rows) and “n” - the number of variables in
#'   the dataframe (number of columns).
#'
#' @param lambda the smoothing parameter; a numeric scalar which takes the
#'   default value of 1600, if unspecified by the user.
#'
#' @return a Txn dataframe, containing the trend data
#'
#' @details The length of the time series should be greater than four and the
#'   value of the smoothing parameter greater than zero for the code to function.
#'   Of course, having a sufficiently long time series is paramount to achieving
#'   meaningful results.
#'
#' @usage hp2(y, lambda = 1600)
#'
#' @keywords hp2 two-sided hp filter gdp macroeconomic smoothing time series
#'
#' @seealso [hp1()]
#'
#' @author Alexandru Monahov, <https://www.alexandrumonahov.eu.org/>
#'
#' @references Balcilar, M. (2019). Miscellaneous Time Series Filters ‘mFilter’. CRAN R Package Library.
#'
#' Drehmann, M., and Yetman, J. (2018). Why You Should Use the Hodrick-Prescott Filter – at Least to Generate Credit Gaps. BIS Working Paper No. 744.
#'
#' Eurostat (2023), Real Gross Domestic Product for European Union (28 countries) [CLVMNACSCAB1GQEU28], National Accounts - GDP.
#'
#' Hamilton, J. D. (2017). ‘Why You Should Never Use the Hodrick-Prescott Filter’. Working Paper Series. National Bureau of Economic Research, May 2017.
#'
#' Hodrick, R. J., and Prescott, E. C. (1997). Postwar U.S. Business Cycles: An Empirical Investigation. Journal of Money, Credit, and Banking 29: 1-16.
#'
#' Hyeongwoo, K. (2004). "Hodrick–Prescott Filter". Notes, Auburn University.
#'
#' Mcelroy, T. (2008). Exact formulas for the Hodrick-Prescott Filter. Econometrics Journal. 11. 209-217.
#'
#' Meyer-Gohde, A. (2010). Matlab code for one-sided HP-filters. QM&RBC Codes 181, Quantitative Macroeconomics & Real Business Cycles.
#'
#' Ravn, M., and Uhlig, H. (2002). On adjusting the Hodrick-Prescott filter for the frequency of observations, The Review of Economics and Statistics 2002; 84 (2): 371–376.
#'
#' Shea, J. (2021). neverhpfilter: An Alternative to the Hodrick-Prescott Filter. CRAN R Package Library.
#'
#' @import Matrix
#'
#' @examples
#' # Generate the data and plot it
#' set.seed(10)
#' y <- as.data.frame(rev(diffinv(rnorm(100)))[1:100])+30
#' colnames(y) <- "gdp"
#' plot(y$gdp, type="l")
#'
#' # Apply the HP filter to the data
#' ytrend = hp2(y)
#' ycycle = y - ytrend
#'
#' # Plot the three resulting series
#' plot(y$gdp, type="l", col="black", lty=1, ylim=c(-10,30))
#' lines(ytrend$gdp, col="#066462")
#' polygon(c(1, seq(ycycle$gdp), length(ycycle$gdp)), c(0, ycycle$gdp, 0), col = "#E0F2F1")
#' legend("bottom", horiz=TRUE, cex=0.75, c("y", "ytrend", "ycycle"), lty = 1,
#'        col = c("black", "#066462", "#75bfbd"))
#'
#' @export

hp2 <- function(y,lambda=1600) {

  ytrend <- matrix(data=NA,nrow=nrow(y),ncol=ncol(y)) # Create an empty matrix of the same size as "y", where the results will be stored
  T = nrow(y)
  n = ncol(y)
  # Preliminary calculations
  x1=rbind(c(1+lambda, -2*lambda, lambda)) # The non-zero elements of the first row of A
  x2=rbind(c(-2*lambda, 1+5*lambda, -4*lambda, lambda)) # The non-zero elements of the second row of A
  x3=rbind(c(lambda, -4*lambda, 1+6*lambda, -4*lambda, lambda)) # The non-zero elements of the j'th row of A, 2<j<T-2
  x2rev=x2[,ncol(x2):1] # Flip columns of matrix # The non-zero elements of the second-to-last row of A
  # x2rev=x2[nrow(x2):1,] # Flip rows of matrix # The non-zero elements of the last row of A
  x1rev=x1[,ncol(x1):1]
  x2rev <- t(as.matrix(x2rev))
  x1rev <- t(as.matrix(x1rev))

  # Make a list (a column vector) containing at position i the row of the i'th non-zero element in A
  I = matrix(3:(T-2), nrow = 1)
  I = do.call(rbind, replicate(5, I, simplify=FALSE)) # |5| (replicate 5 times)
  I = cbind(c(I))
  I = rbind(1,1,1,2,2,2,2,I,T-1,T-1,T-1,T-1,T,T,T)

  J = matrix(1:(T-4), nrow = 1)
  J = do.call(rbind, replicate(5, J, simplify=FALSE)) # |5| (replicate 5 times)
  J = cbind(c(J))

  Temp = matrix(0:4, nrow = 1)
  Temp = t(do.call(rbind, replicate(T-4, Temp, simplify=FALSE)))
  Temp = cbind(c(Temp))

  J = J + Temp
  J = rbind(1,2,3,1,2,3,4,J,T-3,T-2,T-1,T,T-2,T-1,T)

  # Make a list (a column vector) containing at position i the i'th non-zero element in A
  X = t(do.call(rbind, replicate(T-4, x3, simplify=FALSE))) # |5| (replicate 5 times)
  X = cbind(c(X))
  X = rbind(t(x1),t(x2),X,t(x2rev),t(x1rev))

  # Replicate vectors by the number of columns in the input dataframe
  MX = t(do.call(cbind, replicate(ncol(ytrend), X, simplify=FALSE)))
  MX <- t(as.matrix(MX))
  MI = t(do.call(cbind, replicate(ncol(ytrend), I, simplify=FALSE)))
  MI <- t(as.matrix(MI))
  MJ = t(do.call(cbind, replicate(ncol(ytrend), J, simplify=FALSE)))
  MJ <- t(as.matrix(MJ))

  A = sparseMatrix(i=as.vector(MI),j=as.vector(MJ),x=as.vector(MX))
  ytrend <- solve(A, matrix(unlist(y), ncol=ncol(y)), sparse = TRUE)

  ifelse(ncol(y) == 1,
         ytrend <- as.data.frame(ytrend),
         ytrend <- as.data.frame(matrix(ytrend@x, ncol=ncol(y))) * ncol(y) # Store "ytrend" as a tibble
  )
  colnm <- colnames(y) # Get column names from "y"
  colnames(ytrend) <- colnm # Name the columns of "ytrend" so that they have the same names as "y"
  return(ytrend)

}


R/GDPEU.R
#' Real Gross Domestic Product for European Union (28 countries)
#'
#' Units:  Millions of Chained 2010 Euros, Seasonally Adjusted
#'
#' Frequency:  Quarterly
#'
#' Eurostat unit ID: CLV10_MNAC
#' Eurostat item ID = B1GQ
#' Eurostat country ID: EU28
#'
#' Seasonally and calendar adjusted data.
#'
#' For euro area member states, the national currency series are converted into
#' euros using the irrevocably fixed exchange rate. This preserves the same
#' growth rates than for the previous national currency series. Both series
#' coincide for years after accession to the euro area but differ for earlier
#' years due to market exchange rate movements.
#'
#' European Union (28 countries): Belgium, Denmark, Germany, Ireland, Greece,
#' Spain, France, Italy, Luxembourg, the Netherlands, Portugal, the United
#' Kingdom, Austria, Finland, Sweden, Cyprus, the Czech Republic, Estonia,
#' Hungary, Latvia, Lithuania, Malta, Poland, Slovenia, Slovakia, Bulgaria,
#' Romania, and Croatia.
#'
#' Copyright, European Union, http://ec.europa.eu, 1995-2016. Complete terms of
#' use are available at http://ec.europa.eu/geninfo/legal_notices_en.htm#copyright
#'
#' @docType data
#'
#' @usage data(GDPEU)
#'
#' @format A dataframe containing:
#'         **gdp**  the seasonally-adjusted real Gross Domestic Product for 28
#'         European Union countries
#'         **date**  the date of each observation
#'
#' @keywords EU GDP seasonally adjusted quarterly gross domestic product
#'
#' @references Eurostat, Real Gross Domestic Product for European Union (28 countries) [CLVMNACSCAB1GQEU28]
#' (\href{https://ec.europa.eu/eurostat/en/web/main/data}{Eurostat})
#'
#' @source \href{https://fred.stlouisfed.org/series/CLVMNACSCAB1GQEU28}{Data} retrieved from FRED, Federal Reserve Bank of St. Louis.
#'
#' @examples
#' # Load the dataset
#' data(GDPEU)
#'
#' # Plot the y series
#' plot(GDPEU$date, GDPEU$y, type="l")
#'
#' # Remove the date column if not needed and store in a df object
#' df <- GDPEU[,-1]
#'
"GDPEU"


R/hpfilter-package.R
#' @keywords internal
#' @details
#'   hp1 - runs the one-sided HP filter
#'
#'   hp2 - runs the two-sided HP filter
#'
#'   GDPEU - contains a sample GDP dataset
#'
#'   This package implements in R a version of the HP filter very close to that
#'   specified by Meyer-Gohde, A. (2010). Please see the "References" section
#'   for more details.
#'
#' @examples
#' # Generate sample data
#' set.seed(10)
#' y <- as.data.frame(rev(diffinv(rnorm(100)))[1:100])+30
#'
#' # Run hpfilter functions
#' hp1(y, lambda = 400000) # Calculate the one-sided HP filter using a
#'                         # smoothing parameter of 400000
#'
#' hp2(y, lambda = 1600)   # Calculate the two-sided HP filter using a
#'                         # smoothing parameter of 1600
#'
#' @author Alexandru Monahov
#' @references Balcilar, M. (2019). Miscellaneous Time Series Filters ‘mFilter’.
#'  CRAN R Package Library.
#'
#' Drehmann, M., and Yetman, J. (2018). Why You Should Use the Hodrick-Prescott
#'  Filter – at Least to Generate Credit Gaps. BIS Working Paper No. 744.
#'
#' Eurostat (2023), Real Gross Domestic Product for European Union (28 countries)
#'  [CLVMNACSCAB1GQEU28], National Accounts - GDP.
#'
#' Hamilton, J. D. (2017). ‘Why You Should Never Use the Hodrick-Prescott Filter’.
#'  Working Paper Series. National Bureau of Economic Research, May 2017.
#'
#' Hodrick, R. J., and Prescott, E. C. (1997). Postwar U.S. Business Cycles: An
#'  Empirical Investigation. Journal of Money, Credit, and Banking 29: 1-16.
#'
#' Hyeongwoo, K. (2004). "Hodrick–Prescott Filter". Notes, Auburn University.
#'
#' Mcelroy, T. (2008). Exact formulas for the Hodrick-Prescott Filter.
#'  Econometrics Journal. 11. 209-217.
#'
#' Meyer-Gohde, A. (2010). Matlab code for one-sided HP-filters. QM&RBC Codes 181,
#'  Quantitative Macroeconomics & Real Business Cycles.
#'
#' Ravn, M., and Uhlig, H. (2002). On adjusting the Hodrick-Prescott filter for
#'  the frequency of observations, The Review of Economics and Statistics 2002;
#'  84 (2): 371–376.
#'
#' Shea, J. (2021). neverhpfilter: An Alternative to the Hodrick-Prescott Filter.
#'  CRAN R Package Library.
#'
#' @source Copyright Alexandru Monahov, 2023. You may use, modify and redistribute
#'  this code, provided that you give credit to the author and make any derivative
#'  work available to the public for free.
"_PACKAGE"


R/hp1.R
#' One-Sided HP Filter
#'
#' hp1 applies a one-sided Hodrick–Prescott filter derived using the Kalman
#' filter to separate a time series into trend and cyclical components. The
#' smoothing parameter should reflect the cyclical duration or frequency of
#' the data.
#'
#' @param y a dataframe of size Txn, where “T” is the number of observations
#'   for each variable (number of rows) and “n” - the number of variables in
#'   the dataframe (number of columns).
#'
#' @param lambda the smoothing parameter; a numeric scalar which takes the
#'   default value of 1600, if unspecified by the user.
#'
#' @param x_user user defined initial values of the state estimate for each
#'   variable in y. Takes the form of a 2xn matrix. Since the underlying state
#'   vector is 2x1, two values are needed for each variable in y. By default:
#'   if no values are provided, backwards extrapolations based on the first two
#'   observations are used.
#'
#' @param P_user a structural array with n elements, each of which being a 2x2
#'   matrix of initial MSE estimates for each variable in y. By default: if no
#'   values are provided, a matrix with relatively large variances is used.
#'
#' @param discard the number of discard periods, expressed as a numeric scalar.
#'   The user specified amount of values will be discarded from the start of the
#'   sample, resulting in output matrices of size (T-discard)xn. By default: if
#'   no values are provided, is set to 0.
#'
#' @return a (T-discard)xn dataframe, containing the trend data
#'
#' @details The length of the time series should be greater than four and the
#'   value of the smoothing parameter greater than zero for the code to function.
#'   Of course, having a sufficiently long time series is paramount to achieving
#'   meaningful results.
#'
#' @usage hp1(y, lambda = 1600, x_user = NA, P_user = NA, discard = 0)
#'
#' @keywords hp1 one-sided hp filter kalman basel ccyb gap buffer
#'
#' @seealso [hp2()]
#'
#' @author Alexandru Monahov, <https://www.alexandrumonahov.eu.org/>
#'
#' @references Balcilar, M. (2019). Miscellaneous Time Series Filters ‘mFilter’. CRAN R Package Library.
#'
#' Drehmann, M., and Yetman, J. (2018). Why You Should Use the Hodrick-Prescott Filter – at Least to Generate Credit Gaps. BIS Working Paper No. 744.
#'
#' Eurostat (2023), Real Gross Domestic Product for European Union (28 countries) [CLVMNACSCAB1GQEU28], National Accounts - GDP.
#'
#' Hamilton, J. D. (2017). ‘Why You Should Never Use the Hodrick-Prescott Filter’. Working Paper Series. National Bureau of Economic Research, May 2017.
#'
#' Hodrick, R. J., and Prescott, E. C. (1997). Postwar U.S. Business Cycles: An Empirical Investigation. Journal of Money, Credit, and Banking 29: 1-16.
#'
#' Hyeongwoo, K. (2004). "Hodrick–Prescott Filter". Notes, Auburn University.
#'
#' Mcelroy, T. (2008). Exact formulas for the Hodrick-Prescott Filter. Econometrics Journal. 11. 209-217.
#'
#' Meyer-Gohde, A. (2010). Matlab code for one-sided HP-filters. QM&RBC Codes 181, Quantitative Macroeconomics & Real Business Cycles.
#'
#' Ravn, M., and Uhlig, H. (2002). On adjusting the Hodrick-Prescott filter for the frequency of observations, The Review of Economics and Statistics 2002; 84 (2): 371–376.
#'
#' Shea, J. (2021). neverhpfilter: An Alternative to the Hodrick-Prescott Filter. CRAN R Package Library.
#'
#' @examples
#' # Generate the data and plot it
#' set.seed(10)
#' y <- as.data.frame(rev(diffinv(rnorm(100)))[1:100])+30
#' colnames(y) <- "gdp"
#' plot(y$gdp, type="l")
#'
#' # Apply the HP filter to the data
#' ytrend = hp1(y)
#' ycycle = y - ytrend
#'
#' # Plot the three resulting series
#' plot(y$gdp, type="l", col="black", lty=1, ylim=c(-10,30))
#' lines(ytrend$gdp, col="#066462")
#' polygon(c(1, seq(ycycle$gdp), length(ycycle$gdp)), c(0, ycycle$gdp, 0), col = "#E0F2F1")
#' legend("bottom", horiz=TRUE, cex=0.75, c("y", "ytrend", "ycycle"), lty = 1,
#'        col = c("black", "#066462", "#75bfbd"))
#'
#' @export

hp1 <- function(y,lambda=1600,x_user=NA,P_user=NA,discard=0) {

  kalman_update = function(y,F,H,Q,R,j,k,x,P) {
    # Updates the Kalman filter estimation of the state and MSE
    # For details, see Chapter 13 of Hamilton, J.D. (1994). Time Series Analysis. Princeton University Press.
    S = H %*% P %*% t(H) + R
    K = F %*% P %*% t(H)
    K = K / as.double(S) # Kalman gain
    x = F %*% x + K * as.double(y[j,k] - H %*% x) # State estimate
    Temp = F - K %*% H
    P = Temp %*% P %*% t(Temp)
    P = P + Q + K %*% R %*% t(K) # MSE estimate
    return(list("x" = x, "P" = P)) # Store the results to be returned when calling the function as a list (we do this because R functions return one result)
  }

  ytrend <- matrix(data=NA,nrow=nrow(y),ncol=ncol(y)) # Create an empty matrix of the same size as "y", where the Kalman filtration results will be stored
  T = nrow(y)
  n = ncol(y)
  # Kalman preliminaries. The notation follows Chapter 13 of Hamilton, J.D.
  # (1994). Time Series Analysis. with the exception of H, which is equivalent
  # to his H'.
  q = 1 / lambda # the signal-to-noise ration: i.e. var eta_t / var epsilon_t
  F = rbind(c(2,-1),c(1,0)) # The state transition matrix
  H = rbind(c(1,0)) # The observation matrix
  Q = rbind(c(q,0),c(0,0)) # The variance-covariance matrix of the errors in the state equation
  R = 1 # The variance of the error in the observation equation

  for (k in 1:n) { # Run the Kalman filter for each variable
    if (is.na(x_user)) {x = rbind(unlist(c(2*y[1,k]-y[2,k])), unlist(c(3*y[1,k]-2*y[2,k])))} else {x = x_user[,k]} # If the user didn't provide an initial value for state estimate, extrapolate back two periods from the observations
    if (is.na(P_user)) {P = rbind(c(1e5, 0), c(0,1e5))} else {P = P_user[k]} # If the user didn't provide an intial value for the MSE, set a relatively high one
    for (j in 1:T) { # Get the estimates for each period
      klm = kalman_update(y,F,H,Q,R,j,k,x,P) # store the results returned by the function as a list called "klm"
      x = klm$x # get element "x" from the list called "klm" and store it as "x" so that it can be used outside the function kalman_update
      P = klm$P # similarly to x
      ytrend[j,k]=x[2] # The second element of the state is the estimate of the trend
    }
  }

  ycycle = y - ytrend # Deviations from the HP trend

  if (!missing(discard)) { # If the user provided a discard parameter
    if (discard!=0) {
      ytrend = ytrend[-1:-discard,] # Remove the first "discard" periods from the trend series
      ycycle = ycycle[-1:-discard,]
    }
  }

  ytrend <- as.data.frame(ytrend) # Store "ytrend" as a tibble
  colnm <- colnames(y) # Get column names from "y"
  colnames(ytrend) <- colnm # Name the columns of "ytrend" so that they have the same names as "y"
  return(ytrend)

}
